src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,230 @@
+
+`ReturnIfAbrupt` might look like a function call, but it's not. It causes the function where `ReturnIfAbrupt()` occurs to return, not the `ReturnIfAbrupt` function itself. It behaves more like a macro in C-like languages.
+
+`ReturnIfAbrupt` can be used like this:
+
+> 1. Let `obj` be `Foo()` // `obj` is a Completion Record
+> 2. `ReturnIfAbrupt(obj)`
+> 3. `Bar(obj)` // If we're still here, `obj` is the value extracted from the Completion Record
+
+And now the question mark comes into play: `? Foo()` is equivalent to `ReturnIfAbrupt(Foo())`.
+
+Similarly, `Let val be ! Foo()` is equivalent to:
+
+> 1. Let `val` be `Foo()`
+> 2. Assert: `val` is not an abrupt completion
+> 3. Set `val` to `val.[[Value]]`.
+
+[Spec: ReturnIfAbrupt shorthands](https://tc39.es/ecma262/#sec-returnifabrupt-shorthands)
+
+Using this knowledge, we can rewrite `Object.prototype.hasOwnProperty` like this:
+
+> `Object.prototype.hasOwnProperty(P)`
+>
+> 1. Let `P` be `ToPropertyKey(V)`.
+> 2. If `P` is an abrupt completion, return `P`
+> 3. Set `P` to `P.[[Value]]`
+> 4. Let `O` be `ToObject(this value)`.
+> 5. If `O` is an abrupt completion, return `O`
+> 6. Set `O` to `O.[[Value]]`
+> 7. Let `temp` be `HasOwnProperty(O, P)`.
+> 8. If `temp` is an abrupt completion, return `temp`
+> 9. Let `temp` be `temp.[[Value]]`
+> 10. Return `NormalCompletion(temp)`
+
+and `HasOwnProperty` like this:
+
+> `HasOwnProperty(O, P)`
+>
+> 1. Assert: `Type(O)` is `Object`.
+> 2. Assert: `IsPropertyKey(P)` is `true`.
+> 3. Let `desc` be `O.[[GetOwnProperty]](P)`.
+> 4. If `desc` is an abrupt completion, return `desc`
+> 5. Set `desc` to `desc.[[Value]]`
+> 6. If `desc` is `undefined`, return `NormalCompletion(false)`.
+> 7. Return `NormalCompletion(true)`.
+
+We can also rewrite the `[[GetOwnProperty]]` internal method without the exclamation mark like this:
+
+> `O.[[GetOwnProperty]]`
+>
+> 1. Let `temp` be `OrdinaryGetOwnProperty(O, P)`
+> 2. Assert: `temp` is not an abrupt completion
+> 3. Let `temp` be `temp.[[Value]]`
+> 4. Return `NormalCompletion(temp)`
+
+Here we assume that `temp` is a brand new temporary variable which doesn't collide with anything else.
+
+We've also used the knowledge that when a return statement returns something else than a Completion Record, it's implicitly wrapped inside a `NormalCompletion`.
+
+### Side track: Return ? Foo()
+
+The spec uses the notation `Return ? Foo()` - why the question mark?
+
+`Return ? Foo()` expands to:
+
+> 1. Let `temp` be `Foo()`
+> 2. If `temp` is an abrupt completion, return `temp`
+> 3. Set `temp` to `temp.[[Value]]`
+> 4. Return `NormalCompletion(temp)`
+
+Which is the same as `Return Foo()`; it behaves the same way for both abrupt and normal completions.
+
+## Asserts
+
+Asserts in the spec assert invariant conditions of the algorithms. They are added for clarity, but don't add any requirements to the implementation - the implementation doesn't need to check them.
```

Comment:
Formatting nit: I was taught in American schools an em dash be used in situations like these.